File: sdk/basin-js/test/arrow.test.ts
/**
 * Tests for QueryBuilder.toArrow() — Arrow IPC transport.
 *
 * Server contract (crates/basin-rest/src/arrow_ipc.rs):
 *   GET /rest/v1/:table  Accept: application/vnd.apache.arrow.stream
 *   → Content-Type: application/vnd.apache.arrow.stream
 *   → X-Basin-Next-Cursor: <cursor>  (absent when no next page)
 *   → X-Basin-Row-Count: <count>
 *   → Body: Arrow IPC stream
 */

import { tableFromJSON, tableToIPC, type Table } from "apache-arrow";
import { beforeEach, describe, expect, it } from "vitest";
import { createClient } from "../src/index.js";
import { FetchMock } from "./helpers.js";

const ARROW_MIME = "application/vnd.apache.arrow.stream";
const URL_BASE = "http://basin.test";

/** Build an Arrow IPC stream fixture from a list of row objects. */
function makeArrowFixture(rows: Record<string, unknown>[]): Uint8Array {
  const table = tableFromJSON(rows);
  return tableToIPC(table, "stream");
}

/** Wrap Arrow IPC bytes as a Response with the correct content type. */
function arrowResponse(
  bytes: Uint8Array,
  extra: Record<string, string> = {},
  status = 200,
): Response {
  return new Response(bytes, {
    status,
    headers: {
      "content-type": ARROW_MIME,
      ...extra,
    },
  });
}

function makeClient(fetch: FetchMock) {
  return createClient(URL_BASE, "key", { fetch: fetch.impl });
}

describe("QueryBuilder.toArrow()", () => {
  it("sends Accept: application/vnd.apache.arrow.stream header", async () => {
    const fixture = makeArrowFixture([{ id: 1, name: "Alice" }]);
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    await makeClient(fetch).from("orders").toArrow();

    const call = fetch.last;
    expect(call.headers["accept"]).toBe(ARROW_MIME);
    expect(call.method).toBe("GET");
    expect(call.url).toContain("/rest/v1/orders");
  });

  it("forwards query filters on Arrow requests", async () => {
    const fixture = makeArrowFixture([]);
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    await makeClient(fetch).from("orders").eq("status", "open").limit(10).toArrow();

    const url = new URL(fetch.last.url);
    expect(url.searchParams.get("status")).toBe("eq.open");
    expect(url.searchParams.get("limit")).toBe("10");
  });

  it("returns an Arrow Table with the correct column values from IPC response", async () => {
    const rows = [
      { id: 10, name: "Carol" },
      { id: 11, name: "Dave" },
    ];
    const fixture = makeArrowFixture(rows);
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    const table = await makeClient(fetch).from("users").toArrow();

    expect(table.numRows).toBe(2);
    // Column names from the IPC schema.
    const names = table.schema.fields.map((f) => f.name);
    expect(names).toContain("id");
    expect(names).toContain("name");
    // Values.
    const idCol = table.getChild("id");
    expect(idCol?.get(0)).toBe(10);
    expect(idCol?.get(1)).toBe(11);
  });

  it("attaches X-Basin-Next-Cursor to the table as a non-enumerable property", async () => {
    const fixture = makeArrowFixture([{ id: 1 }]);
    const fetch = new FetchMock().queue(
      arrowResponse(fixture, { "x-basin-next-cursor": "cursor-tok" }),
    );
    const table = await makeClient(fetch).from("orders").limit(1).toArrow();
    // The cursor is attached as a non-enumerable own property.
    expect((table as Record<string, unknown>)["x-basin-next-cursor"]).toBe("cursor-tok");
    // It should not appear in JSON serialisation.
    expect(JSON.stringify(Object.keys(table))).not.toContain("x-basin-next-cursor");
  });

  it("falls back to JSON→Arrow when server returns JSON", async () => {
    // Server returns plain JSON (older server / 406 fallback).
    const fetch = new FetchMock().queue(
      new Response(JSON.stringify([{ id: 5, total: 99.5 }]), {
        status: 200,
        headers: { "content-type": "application/json" },
      }),
    );
    const table = await makeClient(fetch).from("orders").toArrow();
    expect(table.numRows).toBe(1);
    const idCol = table.getChild("id");
    // JSON numbers come through as Float64 in the fallback path.
    expect(idCol?.get(0)).toBeDefined();
  });

  it("falls back correctly for paginated JSON response {rows, next_cursor}", async () => {
    const fetch = new FetchMock().queue(
      new Response(
        JSON.stringify({ rows: [{ id: 1 }, { id: 2 }], next_cursor: "c1" }),
        {
          status: 200,
          headers: { "content-type": "application/json" },
        },
      ),
    );
    const table = await makeClient(fetch).from("orders").toArrow();
    expect(table.numRows).toBe(2);
  });

  it("throws BasinApiError on non-2xx responses", async () => {
    const fetch = new FetchMock().queue(
      new Response(JSON.stringify({ code: "E_NOT_FOUND", message: "table not found" }), {
        status: 404,
        headers: { "content-type": "application/json" },
      }),
    );
    await expect(makeClient(fetch).from("missing").toArrow()).rejects.toMatchObject({
      code: "E_NOT_FOUND",
      status: 404,
    });
  });

  it("attaches auth header", async () => {
    const fixture = makeArrowFixture([]);
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    await makeClient(fetch).from("orders").toArrow();
    expect(fetch.last.headers["authorization"]).toBe("Bearer key");
  });

  it("does not attach cursor property when header is absent", async () => {
    const fixture = makeArrowFixture([{ id: 1 }]);
    // No X-Basin-Next-Cursor header.
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    const table = await makeClient(fetch).from("orders").toArrow();
    expect(Object.getOwnPropertyDescriptor(table, "x-basin-next-cursor")).toBeUndefined();
  });

  it("empty IPC stream returns empty table", async () => {
    // Build an IPC stream from empty rows.
    const fixture = makeArrowFixture([]);
    const fetch = new FetchMock().queue(arrowResponse(fixture));
    const table = await makeClient(fetch).from("orders").toArrow();
    // Empty fixture — may have 0 rows.
    expect(table.numRows).toBe(0);
  });
});

File: sdks/go/arrow_test.go
package basin

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/apache/arrow-go/v18/arrow"
	"github.com/apache/arrow-go/v18/arrow/array"
	"github.com/apache/arrow-go/v18/arrow/ipc"
	"github.com/apache/arrow-go/v18/arrow/memory"
)

// ---------------------------------------------------------------------------
// IPC fixture builder
// ---------------------------------------------------------------------------

// buildArrowFixture returns a minimal Arrow IPC stream with one RecordBatch
// containing the given rows. Schema: {id: int64, name: utf8}.
func buildArrowFixture(t *testing.T, rows []struct{ ID int64; Name string }) []byte {
	t.Helper()
	alloc := memory.DefaultAllocator
	schema := arrow.NewSchema([]arrow.Field{
		{Name: "id", Type: arrow.PrimitiveTypes.Int64, Nullable: false},
		{Name: "name", Type: arrow.BinaryTypes.String, Nullable: true},
	}, nil)

	idBldr := array.NewInt64Builder(alloc)
	nameBldr := array.NewStringBuilder(alloc)
	defer idBldr.Release()
	defer nameBldr.Release()

	for _, r := range rows {
		idBldr.Append(r.ID)
		nameBldr.Append(r.Name)
	}
	idCol := idBldr.NewArray()
	nameCol := nameBldr.NewArray()
	defer idCol.Release()
	defer nameCol.Release()

	rec := array.NewRecord(schema, []arrow.Array{idCol, nameCol}, int64(len(rows)))
	defer rec.Release()

	var buf bytes.Buffer
	w := ipc.NewWriter(&buf, ipc.WithSchema(schema), ipc.WithAllocator(alloc))
	if err := w.Write(rec); err != nil {
		t.Fatalf("ipc write record: %v", err)
	}
	if err := w.Close(); err != nil {
		t.Fatalf("ipc write close: %v", err)
	}
	return buf.Bytes()
}

// ---------------------------------------------------------------------------
// decodeArrowIPC
// ---------------------------------------------------------------------------

func TestDecodeArrowIPC(t *testing.T) {
	fixture := buildArrowFixture(t, []struct{ ID int64; Name string }{
		{1, "Alice"},
		{2, "Bob"},
	})

	records, err := decodeArrowIPC(fixture)
	if err != nil {
		t.Fatal(err)
	}
	defer func() {
		for _, r := range records {
			r.Release()
		}
	}()

	if len(records) != 1 {
		t.Fatalf("expected 1 record, got %d", len(records))
	}
	rec := records[0]
	if rec.NumRows() != 2 {
		t.Errorf("NumRows: got %d want 2", rec.NumRows())
	}
	if rec.NumCols() != 2 {
		t.Errorf("NumCols: got %d want 2", rec.NumCols())
	}
	// Verify schema field names.
	if rec.Schema().Field(0).Name != "id" {
		t.Errorf("Field(0).Name: %q", rec.Schema().Field(0).Name)
	}
	if rec.Schema().Field(1).Name != "name" {
		t.Errorf("Field(1).Name: %q", rec.Schema().Field(1).Name)
	}
	// Verify data.
	ids := rec.Column(0).(*array.Int64)
	if ids.Value(0) != 1 || ids.Value(1) != 2 {
		t.Errorf("id values: %d, %d", ids.Value(0), ids.Value(1))
	}
	names := rec.Column(1).(*array.String)
	if names.Value(0) != "Alice" || names.Value(1) != "Bob" {
		t.Errorf("name values: %q, %q", names.Value(0), names.Value(1))
	}
}

func TestDecodeArrowIPCEmpty(t *testing.T) {
	// An IPC stream with schema but zero rows. The writer produces a schema
	// message and an EOS. Decoding yields zero records (the schema message does
	// not count as a RecordBatch).
	//
	// Note: buildArrowFixture with nil rows writes a schema + empty batch + EOS.
	// We build a raw IPC stream with only a schema message and EOS to test the
	// truly-empty case.
	alloc := memory.DefaultAllocator
	schema := arrow.NewSchema([]arrow.Field{
		{Name: "id", Type: arrow.PrimitiveTypes.Int64, Nullable: false},
	}, nil)

	var buf bytes.Buffer
	w := ipc.NewWriter(&buf, ipc.WithSchema(schema), ipc.WithAllocator(alloc))
	// Close without writing any batches → schema + EOS only.
	if err := w.Close(); err != nil {
		t.Fatal(err)
	}

	records, err := decodeArrowIPC(buf.Bytes())
	if err != nil {
		t.Fatal(err)
	}
	for _, r := range records {
		r.Release()
	}
	if len(records) != 0 {
		t.Errorf("expected 0 records for schema-only IPC stream, got %d", len(records))
	}
}

// ---------------------------------------------------------------------------
// QueryBuilder.Arrow — HTTP mock transport
// ---------------------------------------------------------------------------

func TestQueryBuilderArrowNativePath(t *testing.T) {
	fixture := buildArrowFixture(t, []struct{ ID int64; Name string }{
		{10, "Carol"},
		{11, "Dave"},
	})

	mux := http.NewServeMux()
	mux.HandleFunc("/rest/v1/users", func(w http.ResponseWriter, r *http.Request) {
		// Verify the Accept header.
		if r.Header.Get("Accept") != arrowMIME {
			http.Error(w, "expected Accept: "+arrowMIME, 406)
			return
		}
		w.Header().Set("Content-Type", arrowMIME)
		w.Header().Set("X-Basin-Next-Cursor", "cursor-tok")
		w.Header().Set("X-Basin-Row-Count", "2")
		w.WriteHeader(200)
		_, _ = w.Write(fixture)
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := New(srv.URL, "test-api-key", WithProjectID("proj-01"))
	result, err := client.Table("users").Arrow(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	defer result.Release()

	if len(result.Records) != 1 {
		t.Fatalf("expected 1 record, got %d", len(result.Records))
	}
	if result.Records[0].NumRows() != 2 {
		t.Errorf("NumRows: got %d want 2", result.Records[0].NumRows())
	}
	if result.NextCursor != "cursor-tok" {
		t.Errorf("NextCursor: got %q want cursor-tok", result.NextCursor)
	}
	if result.RowCount != 2 {
		t.Errorf("RowCount: got %d want 2", result.RowCount)
	}
}

func TestQueryBuilderArrowFallbackToJSON(t *testing.T) {
	// Server returns JSON (older server / fallback path).
	mux := http.NewServeMux()
	mux.HandleFunc("/rest/v1/orders", func(w http.ResponseWriter, r *http.Request) {
		// Ignore the Accept header; return JSON.
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(200)
		_ = json.NewEncoder(w).Encode([]map[string]any{
			{"id": 1, "total": 99.5},
			{"id": 2, "total": 45.0},
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := New(srv.URL, "test-api-key", WithProjectID("proj-01"))
	result, err := client.Table("orders").Arrow(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	defer result.Release()

	// In the fallback path, Records is nil and FallbackRows is populated.
	if result.Records != nil {
		t.Errorf("expected Records to be nil in fallback path")
	}
	if len(result.FallbackRows) != 2 {
		t.Errorf("FallbackRows len: got %d want 2", len(result.FallbackRows))
	}
}

func TestQueryBuilderArrowAuthHeader(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/rest/v1/orders", func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer test-key" {
			http.Error(w, "missing auth", 401)
			return
		}
		fixture := buildArrowFixture(t, nil)
		w.Header().Set("Content-Type", arrowMIME)
		w.WriteHeader(200)
		_, _ = w.Write(fixture)
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := New(srv.URL, "test-key")
	result, err := client.Table("orders").Arrow(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	result.Release()
}

func TestQueryBuilderArrowWithFilters(t *testing.T) {
	// Verify that query params are forwarded on Arrow requests.
	var gotQuery string
	mux := http.NewServeMux()
	mux.HandleFunc("/rest/v1/orders", func(w http.ResponseWriter, r *http.Request) {
		gotQuery = r.URL.RawQuery
		fixture := buildArrowFixture(t, nil)
		w.Header().Set("Content-Type", arrowMIME)
		w.WriteHeader(200)
		_, _ = w.Write(fixture)
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := New(srv.URL, "key")
	result, err := client.Table("orders").Eq("status", "open").Limit(10).Arrow(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	result.Release()

	if gotQuery == "" {
		t.Error("expected query params, got empty string")
	}
	// Should contain status=eq.open and limit=10.
	if !containsParam(gotQuery, "status", "eq.open") {
		t.Errorf("query params missing status filter: %q", gotQuery)
	}
	if !containsParam(gotQuery, "limit", "10") {
		t.Errorf("query params missing limit: %q", gotQuery)
	}
}

func TestQueryBuilderArrowErrorResponse(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/rest/v1/orders", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(404)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code":    "E_NOT_FOUND",
			"message": "table not found",
		})
	})
	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := New(srv.URL, "key")
	_, err := client.Table("orders").Arrow(context.Background())
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	be, ok := err.(*BasinError)
	if !ok {
		t.Fatalf("expected *BasinError, got %T: %v", err, err)
	}
	if be.Code != "E_NOT_FOUND" {
		t.Errorf("Code: got %q want E_NOT_FOUND", be.Code)
	}
}

// ---------------------------------------------------------------------------
// arrowIsNative / arrowParseRowCount helpers
// ---------------------------------------------------------------------------

func TestArrowIsNative(t *testing.T) {
	cases := []struct {
		ct   string
		want bool
	}{
		{arrowMIME, true},
		{arrowMIME + "; charset=utf-8", true},
		{"application/json", false},
		{"", false},
	}
	for _, tc := range cases {
		got := arrowIsNative(tc.ct)
		if got != tc.want {
			t.Errorf("arrowIsNative(%q): got %v want %v", tc.ct, got, tc.want)
		}
	}
}

func TestArrowParseRowCount(t *testing.T) {
	cases := []struct {
		s    string
		want int64
	}{
		{"42", 42},
		{"0", 0},
		{"", 0},
		{"abc", 0},
	}
	for _, tc := range cases {
		if got := arrowParseRowCount(tc.s); got != tc.want {
			t.Errorf("arrowParseRowCount(%q): got %d want %d", tc.s, got, tc.want)
		}
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func containsParam(rawQuery, key, value string) bool {
	for _, pair := range splitParams(rawQuery) {
		if pair[0] == key && pair[1] == value {
			return true
		}
	}
	return false
}

func splitParams(q string) [][2]string {
	var out [][2]string
	for _, pair := range bytes.Split([]byte(q), []byte("&")) {
		kv := bytes.SplitN(pair, []byte("="), 2)
		if len(kv) == 2 {
			// URL-decode the value for comparison.
			val := string(kv[1])
			val = urlDecode(val)
			out = append(out, [2]string{string(kv[0]), val})
		}
	}
	return out
}

func urlDecode(s string) string {
	out := make([]byte, 0, len(s))
	for i := 0; i < len(s); i++ {
		if s[i] == '%' && i+2 < len(s) {
			var b byte
			for _, c := range []byte(s[i+1 : i+3]) {
				b <<= 4
				if c >= '0' && c <= '9' {
					b |= c - '0'
				} else if c >= 'a' && c <= 'f' {
					b |= c - 'a' + 10
				} else if c >= 'A' && c <= 'F' {
					b |= c - 'A' + 10
				}
			}
			out = append(out, b)
			i += 2
		} else if s[i] == '+' {
			out = append(out, ' ')
		} else {
			out = append(out, s[i])
		}
	}
	return string(out)
}
